index.js

```diff
@@ -407,56 +404,106 @@ class K8sExecutor extends Executor {
             if (resp.statusCode !== 201) {
                 throw new Error(`Failed to create pod:${JSON.stringify(resp.body)}`);
             }
-            const podName = resp.body.metadata.name;
-            let statusOptions = {
-                uri: `${this.podsUrl}/${podName}/status`,
-                method: 'GET',
-                headers: { Authorization: `Bearer ${this.token}` },
-                strictSSL: false,
-                maxAttempts: this.maxAttempts,
-                retryDelay: this.retryDelay,
-                retryStrategy: this.scheduleStatusRetryStrategy,
-                json: true
-            };
-            let res = await this.breaker.runCommand(statusOptions);
-
-            this.checkPodResponse(res, buildId);
+            const { status, nodeName, podName, waitingReason } = await this.getPodStatus({
+                podName: resp.body.metadata.name,
+                buildId,
+                retryStrategyFn: this.scheduleStatusRetryStrategy
+            });
+            const isPodInitializing = this.checkPodResponse({ status, waitingReason, buildId, podName });
             const updateConfig = {
                 apiUri: this.ecosystem.api,
                 buildId,
                 token
             };
 
-            if (res.body.spec && res.body.spec.nodeName) {
-                updateConfig.stats = {
-                    hostname: res.body.spec.nodeName,
-                    imagePullStartTime: new Date().toISOString()
-                };
+            if (nodeName) {
+                Object.assign(updateConfig, {
+                    stats: {
+                        hostname: nodeName,
+                        imagePullStartTime: new Date().toISOString()
+                    }
+                });
             } else {
-                updateConfig.statusMessage = 'Waiting for resources to be available.';
+                Object.assign(updateConfig, { statusMessage: 'Waiting for resources to be available.' });
             }
+
             await this.updateBuild(updateConfig);
-            await new Promise(r => setTimeout(r, this.podStatusQueryDelay));
-            statusOptions = {
-                uri: `${this.podsUrl}/${podName}/status`,
-                method: 'GET',
-                headers: { Authorization: `Bearer ${this.token}` },
-                strictSSL: false,
-                maxAttempts: this.maxAttempts,
-                retryDelay: this.retryDelay,
-                retryStrategy: this.pendingStatusRetryStrategy,
-                json: true
-            };
-            res = await this.breaker.runCommand(statusOptions);
-            this.checkPodResponse(res, buildId);
 
-            return null;
+            return (
+                isPodInitializing &&
+                (await this.callWithRetry(
```

Comment:
does this result in worker waiting for pod to start running ? What if it's pulling a 10G image ? Is it going to wait until the image pull is successfully complete ?

Reply:
it always did, rabbitmq needs to either ack or nack every msg it sends. However, it doesn't affect the processing of other messages. All messages are in a separate thread

Reply:
It will affect the processing speed. we pull like 20 msgs and how fast we ack or nack, if ack or nack gets delayed, msgs will process slowly, and eventually backlog will increase.

Reply:
ok, I see. Currently, we only get the pod waitingReason as PodInitializing after all the existing retries have exhausted. I want to test this in beta to see if we can fetch the actual reason ImagePullBackOff for the correct pod. Also have added logs to get the root cause. I will get rid of the additional retry once I can pinpoint the status issue.